protocol/chainlib/provider_node_subscription_manager.go

```diff
@@ -0,0 +1,520 @@
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unsupported API interface: %s", apiCollection.GetCollectionData().ApiInterface)
+	}
+
+	marshalledMsg, err := gojson.Marshal(convertedMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return marshalledMsg, nil
+}
+
+func (pnsm *ProviderNodeSubscriptionManager) signReply(ctx context.Context, reply *pairingtypes.RelayReply, consumerAddr sdk.AccAddress, chainMessage ChainMessage, request *pairingtypes.RelayRequest) error {
+	// Send the first setup message to the consumer in a go routine because the blocking listening for this channel happens after this function
+	dataReliabilityEnabled, _ := pnsm.chainParser.DataReliabilityParams()
+	blockLagForQosSync, averageBlockTime, blockDistanceToFinalization, blocksInFinalizationData := pnsm.chainParser.ChainBlockStats()
+	relayTimeout := GetRelayTimeout(chainMessage, averageBlockTime)
+
+	if dataReliabilityEnabled {
+		var err error
+		latestBlock, _, requestedHashes, modifiedReqBlock, _, updatedChainMessage, err := pnsm.relayFinalizationBlocksHandler.GetParametersForRelayDataReliability(ctx, request, chainMessage, relayTimeout, blockLagForQosSync, averageBlockTime, blockDistanceToFinalization, blocksInFinalizationData)
+		if err != nil {
+			return err
+		}
+
+		err = pnsm.relayFinalizationBlocksHandler.BuildRelayFinalizedBlockHashes(ctx, request, reply, latestBlock, requestedHashes, updatedChainMessage, relayTimeout, averageBlockTime, blockDistanceToFinalization, blocksInFinalizationData, modifiedReqBlock)
+		if err != nil {
+			return err
+		}
+	}
+
+	var ignoredMetadata []pairingtypes.Metadata
+	reply.Metadata, _, ignoredMetadata = pnsm.chainParser.HandleHeaders(reply.Metadata, chainMessage.GetApiCollection(), spectypes.Header_pass_reply)
+	reply, err := lavaprotocol.SignRelayResponse(consumerAddr, *request, pnsm.privKey, reply, dataReliabilityEnabled)
+	if err != nil {
+		return err
+	}
+	reply.Metadata = append(reply.Metadata, ignoredMetadata...) // appended here only after signing
+	return nil
+}
+
+func (pnsm *ProviderNodeSubscriptionManager) handleNewNodeMessage(ctx context.Context, hashedParams string, nodeMsg interface{}) {
+	pnsm.lock.RLock()
+	defer pnsm.lock.RUnlock()
+
+	if _, ok := pnsm.activeSubscriptions[hashedParams]; !ok {
+		utils.LavaFormatDebug("No hashed params in handleNewNodeMessage, connection might have been closed", utils.LogAttr("hash", hashedParams))
+		return
+	}
+	// Sending message to all connected consumers
+	for consumerAddrString, connectedConsumerContainer := range pnsm.activeSubscriptions[hashedParams].connectedConsumers {
+		utils.LavaFormatTrace("ProviderNodeSubscriptionManager:startListeningForSubscription() sending to consumer", utils.LogAttr("consumerAddr", consumerAddrString), utils.LogAttr("hashedParams", utils.ToHexString(hashedParams)))
+
+		copiedRequest := &pairingtypes.RelayRequest{}
+		copyRequestErr := protocopy.DeepCopyProtoObject(connectedConsumerContainer.firstSetupRequest, copiedRequest)
```

Comment:
every node message we copy the firstSetup Request, and parse it, it doesn't change, and does not depend on nodeMsg, we can do it once, save it in the parent and save compute, subscriptions are heavily used, let's save needless actions where we can, also no need to copy just to sign the Rlock is enough

Reply:
I rather not waste time on this small enhancement. lets leave it to a later day for optimizations. I will add a comment so we dont forget.